Step5 리뷰 요청 드립니다 :)

kwanguuuu opened the pull request:
step5 마지막이네요. ㅎㅎ
마지막까지 잘 부탁드리겠습니다.
f9129670e310cae1655fc2368578c2a4e0ac99df

이 커밋에 대해서 같이 얘기해봤으면 해요..!

kwanguuuu committed:
[step4-pr] chore: 목적없는 테스트 코드 삭제 - pr 리뷰 반영

- 리뷰반영

kwanguuuu committed:
[step4-pr] chore: parameterized 테스트로 변경

kwanguuuu committed:
[step4-pr] chore: 테스트명 변경

- 리뷰 반영 : https://github.com/next-step/kotlin-racingcar/pull/539#discussion_r873760231

kwanguuuu committed:
[step5] refactor: domain, controller,dto 패키지 생성, 관련 파일 위치 이동

kwanguuuu committed:
[step5] refactor: MoveResults의 farthestDistanceCarNames 삭제, 테스트 코드 삭제

kwanguuuu committed:
[step5] fix: Drivers.kt에 getWinnerResults() 추가. 테스트작성

- help: 테스트 검증을 어떻게 하는게 좋을지 잘 작성되지 않는데.. 같이 얘기해보면 좋을것 같아요

kwanguuuu committed:
[step5] chore: 테스트용 AlwaysFalseMoveStrategy.kt, TestNameStrategy.kt 생성 & 적용,
프로덕션 코드 strategy 패키지로 관련코드 이동

kwanguuuu committed:
[step5] chore: 테스트 코드 패키지 프로덕션이랑 같은 뎊스로 이

kwanguuuu committed:
[step5] refactor: ResultView의 레이싱결과가 한번만 호출되도록 수정

- 리뷰 반영 :https://github.com/next-step/kotlin-racingcar/pull/539/files/c10debf193b1eb23e528dace73153567ccca6240#r873745069

kwanguuuu committed:
[step5] refactor: Drivers의 NameLengthLimitStrategy() default 값 세팅 삭제

wisemuji commented on src/main/kotlin/racingcar/controller/RacingGame.kt:
```diff
@@ -1,27 +1,28 @@
 
 class RacingGame(carNames: List<String>, trialCount: Int, moveStrategy: MoveStrategy, nameStrategy: NameStrategy) {
     private val drivers: Drivers = Drivers(carNames, moveStrategy, nameStrategy)
+    private var _results: MutableList<MoveResults> = mutableListOf()
```
variable이 아니어도 될 것 같은데 어떻게 생각하시나요?

kwanguuuu replied:
참조하는 mutableList가 다른 mutableList로 변경될 일은 없으니 val로 변경하는게 맞는것 같습니다 ㅎ

wisemuji commented on src/main/kotlin/racingcar/controller/RacingGame.kt:
```diff
@@ -1,27 +1,28 @@
     private var _trialCount: Int = trialCount
 
     val trialCount get() = _trialCount
+    val results: List<MoveResults> get() = _results
```
```suggestion
    val results: List<MoveResults> get() = _results.toList()
```
방어적 복사를 사용하시면 좋습니다! .toList()를 안해주면 results를 외부에서 mutable로 캐스팅 했을 때 내부에 숨겨둔 _results를 수정할 수 있어요.

kwanguuuu replied:
오 넵! toList 변경하겠습니다 :) 감사합니다

wisemuji commented on src/main/kotlin/racingcar/Main.kt:
```diff
@@ -16,7 +16,7 @@ fun main() {
     repeat(racingGame.trialCount) {
         racingGame.play()
-        ResultView.showRecordsAsDash(racingGame.getPlayResults())
     }
```
ResultView의 호출이 최소가 되도록 잘 구현해주셨어요!!

하나 더 고민해보자면, racingGame을 만들 때 이미 trialCount를 넘기고 있는데 이 repeat 하는 역할도 main이 아닌 racingGame이 들고 있도록 하면 어때요? racingGame에게 메세지만 전달해보세요.

kwanguuuu replied:
https://github.com/next-step/kotlin-racingcar/pull/617/commits/68c042c5750942c9e841dc4287ed0f37c125a29a

해당 커밋에 반영했습니다.

repeat을 내부로 가져오니 불필요한 프로퍼티&객체가 발생하네요
`var _trialCount` 랑 `GameOverException`을 삭제하도록 하겠습니다.

GameOverException 같은 경우엔, 제가 repeat이 밖에서 수행되기 때문에 방어적으로 작성한 예외인데,

요구사항에서도 이걸 구현하란건 없었고, 오히려 없어지는게 더 명확한 코드가 된 것 같아요.!

wisemuji commented on src/main/kotlin/racingcar/view/ResultView.kt:
```diff
@@ -22,11 +22,17 @@ object ResultView {
 
-    fun showRecordsAsDash(results: MoveResults) {
-        results.result.map { result ->
+    fun showAllRacingRecords(results: List<MoveResults>) {
+        results.forEach { eachResult ->
+            showEachRecord(eachResult)
+            println()
+        }
+    }
+
+    private fun showEachRecord(moveResults: MoveResults) {
+        moveResults.data.map { result ->
             showResultView(result.carName, getDashView(result.moveDistance))
         }
```
```suggestion
    fun showAllRacingRecords(results: List<MoveResults>) {
        results.forEach(::showEachRecord)
    }

    private fun showEachRecord(moveResults: MoveResults) {
        moveResults.data.map { result ->
            showResultView(result.carName, getDashView(result.moveDistance))
        }
        println()
```

[method reference](https://proandroiddev.com/kotlin-lambda-vs-method-reference-fdbd175f6845)를 활용하면 arguments로 전달되는 값이 수정되지 않았음을 보장할 수 있다는 장점이 있어요. 이번 기회에 사용법을 익혀보시면 어떨까요?

kwanguuuu replied:
리뷰 반영했습니다 :)

wisemuji commented on src/test/kotlin/racingcar/domain/DriversTest.kt:
```diff
@@ -53,11 +56,8 @@ class DriversTest {
         // given
         val alwayMoveDriver = Drivers(
```
변수 네이밍도 움직일 수 없는 차임을 알 수 있는 값으로 설정해주세요!

kwanguuuu replied:
리뷰 반영 했습니다 !

wisemuji commented on src/test/kotlin/racingcar/domain/strategy/TestNameStrategy.kt:
```diff
@@ -0,0 +1,7 @@
+package racingcar.domain.strategy
+
+class TestNameStrategy : NameStrategy {
```
테스트 코드에서 TestNameStrategy를 사용하는 곳과 NameLengthLimitStrategy를 사용하는 곳은 무슨 차이가 있나요?

kwanguuuu replied:
TestNameStrategy 내부에서 아무것도 밸리데이션을 하지 않기에,
테스트코드에서 NameLengthLimitStrategy를 참조한다면, 테스트 실패가 발생할 수 있을것 같아 테스트 클래스를 작성했습니다.

wisemuji commented on src/main/kotlin/racingcar/domain/Drivers.kt:
```diff
@@ -20,6 +20,13 @@ class Drivers(
             .forEach { car -> car.move() }
     }
 
+    fun getWinnerResults(): List<MoveResults.MoveResult> {
```
List<MoveResults.MoveResult>이 아닌 MoveResults가 되어도 되지 않을까요?

kwanguuuu replied:
MoveResult를 반환하는게 더 좋을것 같아요. 일급객체를 만들었으니. 일급객체를 사용해야겠습니다..

wisemuji commented on src/main/kotlin/racingcar/controller/RacingGame.kt:
```diff
@@ -31,7 +32,8 @@ class RacingGame(carNames: List<String>, trialCount: Int, moveStrategy: MoveStra
-        return drivers.getMoveResults().farthestDistanceCarNames
+        return drivers.getWinnerResults()
+            .map { it.carName }
     }
```
저는 이 로직은 뷰의 요구사항이었기 때문에 뷰가 들고 있어도 된다고 생각하긴 합니다! 광유님은 어떻게 생각하세요?

kwanguuuu replied:
작성 했을땐 RacingGame이 데이터들을 조작한다고 생각했어요.! (ex. play, getWinner)
순수한 결과로써의 데이터는 Drivers, Car가 가지고 있구요.

그래서 RacingGame이 drivers의 winnerResults를 호출하고,
그 값으로 winners 즉 저는 winners는 우승자라고 생각을 해서 우승자만 반환하는 로직을 여기다 작성했는데요..!

---
수현님 의견은 아래 내용과 비슷할까요??
- RacingGame이 getWinners를 통해서 drivers의 winnerResults를 호출한 것만으로도 제 역할을 한 것이다.

패싱한 것만으로도 역할이 있다면, 이름을 찾는 로직은 뷰에 있는것도 어울리는 것 같아요.

wisemuji replied:
네 맞습니다! 
이렇게 생각해보시면 더 이해하기 쉬우실 것 같아요. 
만약에 뷰의 요구사항이 우승자의 이름을 capitalize하도록 요구사항이 바뀐다고 생각해본다면 이름을 리턴하는 로직이 도메인에 있기 때문에 capitalize 하는 역할도 도메인에 있어야 하나? 고민될 수 있어요. 만약 도메인에 둔다고 하면 뷰의 수정사항에 도메인까지 영향을 주게 만들기도 하구요. 
하지만 도메인과 뷰의 요구사항을 철저하게 분리한다면 뷰의 요구사항이 바뀐다고 하더라도 도메인 코드에는 영향을 주지 않습니다.(참고로 이는 클린 아키텍쳐의 아주아주 기본적인 아이디어 중 하나입니다!)

wisemuji commented on src/main/kotlin/racingcar/domain/Drivers.kt:
```diff
@@ -20,6 +20,13 @@ class Drivers(
+    fun getWinnerResults(): List<MoveResults.MoveResult> {
+        val farthestDistance = cars.maxOf { it.moveResult.moveDistance }
+        return cars
+            .map { it.moveResult }
```
요기서는 getMoveResults()를 호출해도 되지 않을까요?

kwanguuuu replied:
제가 코드를 수정하다보니 현재
getMoveResult()는 driver에만 있는데요

getMoveResults가 `cars.map { it.moveResult }` 만 반환하는게 어떻겠냐는 의견이셨을까요??

```kotlin
    fun getMoveResults(): List<MoveResults.MoveResult> {
        return cars.map { it.moveResult }
    }
```

wisemuji replied:
기존의 getWinnerResults()에서 MoveResults로 매핑하는 부분을 getMoveResults() 호출로 대체할 수도 있지 않을까 해서 여쭤봤던거였어요! 새로 바뀐 diff를 보니 이 구조에서는 어렵겠군요 🤔

wisemuji reviewed (changes requested):
5단계 구현 잘 해주셨습니다 광유님 👏  4단계의 리뷰까지 꼼꼼하게 잘 반영해주셨군요!!

> 테스트 검증을 어떻게 하는게 좋을지 잘 작성되지 않는데.. 같이 얘기해보면 좋을것 같아요

질문을 하나 남겨주셨는데요, 제 생각에는 다음 두 로직이 검증되고 있지 않은 것 같아요.
- Drivers의 getWinnerResults()에서는 실제 경기에서 우승한 우승자를 반환하고 있나? 
  - 이동결과에 대해서는 확인하고 있지만 실제 우승한 우승자인지 알 수 없습니다.
- RacingGame은 우승자를 반환하나? (==Drivers의 getWinnerResults()를 호출하고 있나?)

추가로 제가 남긴 코멘트중에서 우승자의 이름을 반환하는 getWinners()가 RacingGame에 있을 필요가 있을지도 같이 고민해보시면 좋을 것 같습니다.

코멘트를 몇 가지 남겨두었으니 충분히 고민해보시고 말씀해주세요! 궁금하거나 더 이야기하고 싶은 내용은 언제든 DM 주시면 됩니다. 🙂 
이제 정말 끝이 보이네요!!!! 🌈

kwanguuuu committed:
[step5] refactor: RacingGame play 수정

- play 수정하면서 관련된 객체 내부 variable 프로퍼티 삭제
- 리뷰반영 : https://github.com/next-step/kotlin-racingcar/pull/617#discussion_r876026605

kwanguuuu committed:
[step5] refactor: 메서드 레퍼런스 활용 (리뷰 반영)

kwanguuuu committed:
[step5] refactor: 테스트 조건이 변경돼, 실패하는 테스트 삭제, 미사용 클래스 삭제(GameOverException)

kwanguuuu committed:
[step5] refactor: getWinners 반환타입 변경, 의존하던 함수 수정

wisemuji commented on src/main/kotlin/racingcar/controller/RacingGame.kt:
```diff
@@ -0,0 +1,32 @@
+        }
+    }
+
+    fun isPlayable(): Boolean {
+        return trialCount > GAME_END_COUNT
+    }
```
이 코드는 outdated 된 것 같습니다 🙂

kwanguuuu replied:
삭제하겠습니다.!

wisemuji commented on src/test/kotlin/racingcar/controller/RacingGameTest.kt:
```diff
@@ -0,0 +1,28 @@
+import racingcar.domain.strategy.NameStrategy
+import racingcar.domain.strategy.TestNameStrategy
+
+class RacingGameTest {
```
RacingGameTest에서 테스트하고 있지 않은 로직들이 있는 것 같은데 이를 검증해보면 어떨까요?

wisemuji commented on src/main/kotlin/racingcar/domain/Drivers.kt:
```diff
@@ -0,0 +1,35 @@
+    }
+
+    fun getMoveResults(): MoveResults {
+        cars.map { it.moveResult }
```
이 코드는 유효하지 않은 코드로 보입니다.

wisemuji commented on src/main/kotlin/racingcar/domain/Position.kt:
```diff
@@ -1,4 +1,4 @@
-package racingcar.car
+package racingcar.domain
 
 enum class Position(desc: String) {
```
Position의 desc 필드와 STOP 값은 사용되지 않는 것 같아요!

kwanguuuu replied:
초기 설계엔 이동에 대한 정의를 
이동명령은 항상 내리고, 이동/이동안함의 결과를 저장하려고 했는데, ( 그래서 POSITION을 생성 )

리팩토링 하다보니, 이동한 경우에만 움직임을 기록하면 되니 Position을 아예 삭제하는 쪽으로 개선해보겠습니다.

wisemuji commented on src/main/kotlin/racingcar/view/ResultView.kt:
```diff
@@ -1,14 +1,14 @@
-        winners.toString()
-        println("${toWinnerView(winners)}가 최종 우승했습니다.")
+    fun showRaceWinner(winners: MoveResults) {
+        val winnerNames = winners.data.map { it.carName }
```
리뷰 잘 반영해주셨습니다 👍 👍 👍

wisemuji commented on src/test/kotlin/racingcar/domain/DriversTest.kt:
```diff
@@ -43,34 +46,49 @@ class DriversTest {
     }
+
+    @Test
+    fun `우승자의 이동결과를 반환한다`() {
```
게더에서 이야기했던 것처럼 우승자에 대한 테스트가 구체적으로 정의되면 더욱 좋을 것 같아요! [여기](https://github.com/next-step/kotlin-racingcar/pull/617#pullrequestreview-977152568)서 제가 남긴 코멘트 참고 부탁드립니다 🙏

kwanguuuu replied:
@wisemuji 
수현님 이 부분 시간이 지나서 제가 좀 놓친거 같아요.!
근데 지금 테스트 코드를 보면,

`우승자에 대한 테스트`는 RacingGame으로 가야 하지 않을까요.!? 
RacingGame에서 우승자를 반환한 것이, Drivers가 반환하는 우승자와 동일함을 증빙하면 테스트 성공이 되는것 같은데,
이게 수현님이 의도한 건지 잘 모르겠어요 :(....
https://github.com/next-step/kotlin-racingcar/pull/617/commits/33a6c1317987deffad67c1365774cc6f65ea9957

제가 이부분은 뭔가 명확하게 아직 이해하지 못하고 있는거 같아요.. (마지막 남은 과제 같은..)

wisemuji replied:
> 우승자에 대한 테스트는 RacingGame으로 가야 하지 않을까요.!?

getWinnerResults() 함수를 통해 가장 멀리 간 자동차를 계산하고, 리턴하는 역할은 Drivers가 하고 있기 때문에 만약 우승자 로직에 문제가 있다면 DriversTest에서 검증되어야 하지 않을까요? 만약 RacingGameTest에 해당 로직이 가게 된다면 RacingGameTest가 깨졌음에도 Drivers 클래스가 수정되는 어색한 상황이 벌어지게 됩니다.

> RacingGame에서 우승자를 반환한 것이, Drivers가 반환하는 우승자와 동일함을 증빙하면 테스트 성공이 되는것 같은데,
이게 수현님이 의도한 건지 잘 모르겠어요 :(....

제가 전에 코멘트로 다음 로직이 검증되고 있지 않는 것 같다고 남겼었는데요,

* Drivers의 getWinnerResults()에서는 실제 경기에서 우승한 우승자를 반환하고 있나?
  * 이동결과에 대해서는 확인하고 있지만 실제 우승한 우승자인지 알 수 없습니다.
* RacingGame은 우승자를 반환하나? (==Drivers의 getWinnerResults()를 호출하고 있나?)

이미 실제 경기에서 우승한 우승자를 반환 테스트는 DriversTest에서 하고 있기 때문에 RacingGame이 Drivers의 getWinnerResults()를 호출하고 있는지만 검증이 된다면 그 방법은 크게 상관없다고 생각합니다. RacingGameTest에서도 동일하게 실제 경기에서 우승한 우승자를 반환하고 있는지를 검증해도 괜찮은 것 같아요. 어쨌든 RacingGame.getWinners()를 통해 받아오기를 기대하는 것도 우승자를 리턴하는 것이라 해당 로직을 다시 검증하는 것도 충분히 자연스럽다고 생각해요!

@kwanguuuu 충분히 고민해보시고 좀 더 이야기하고싶으시면 언제든 말씀해주세요 ㅎㅎ

wisemuji reviewed (changes requested):
안녕하세요 광유님!

리뷰 잘 반영해주셨어요 👏  코드가 정말 깔끔해진게 눈에 보입니다! 💯 
아마 이번 코멘트만 반영되고 나면 제가 어프룹/머지하게 될 것 같습니다 ☺️ 
마지막까지 조금만 더 화이팅이에요!

궁금한 점이나 논의하고 싶은 내용이 있으면 언제든 DM 주세요! 🙏 💪

kwanguuuu committed:
[step5] refactor: RacingGame.isPlayable()삭제 - outdated, 테스트코드 함께 삭제

kwanguuuu committed:
[step5] fix: 불필요하게 호출되는 코드라인 삭제

kwanguuuu committed:
[step5] refactor: Position enum 삭제, 관련 프로퍼티 수정

kwanguuuu committed:
[step5] test: 레이싱 게임 우승자와 드라이버 우승자가 동일함 증빙 테스트

wisemuji reviewed (changes requested):
리뷰 반영하시느라 고생이 많으십니다 🙇 

따로 말씀드린 것처럼 다음 코멘트에 대해 한번 고민해보세요! 다음 요청에는 머지하겠습니다 ㅎㅎ

https://github.com/next-step/kotlin-racingcar/pull/617#discussion_r879565039

kwanguuuu committed:
[step5] refactor: 레이싱게임,드라이버 클래스 리팩토링

kwanguuuu commented:
수현님 안녕하세요! 일주일 만에 리뷰를 남깁니닷. 하하하

> RacingGame은 우승자를 반환하나? (==Drivers의 getWinnerResults()를 호출하고 있나?)

위의 내용을 어떻게 테스트로 증빙할지 고민이 됐는데요,
결론부터 말하면, RacingGame에 Drivers를 주입받는 방식으로 코드를 리팩토링해 수정했습니다.
이유는 기존 RacingGame은 Driver 클래스를 위한 인자들을 외부에서 주입받고, 내부에서 드라이버를 재생성 하는 방식으로 작성했는데요

이렇게 되니 RacingGame이 가지고 있는 드라이버 클래스가, 레이싱게임의 함수호출 결과로 변환된 값인지 아닌지 비교를 할 수가 없는 구조가 돼서요.

racingGame가 drivers의 getWinnerResults를 호출하는지는,
외부의 driver가 직접 호출한 getWinnerResults와 
RacingGame이 호출한 getWinners가 같은 객체를 조작하고 있음을 증빙해야 알 수 있는것 같더라구요.

그래서 RacingGame의 생성시 Drvier를 받고,
Drvier의 driveAll 호출은 RacingGame이 가지고 있는 전략을 사용하도록 코드를 일부 수정했습니다.

위와 같은 방식으로 증빙 할 수 있을것 같은데 의도한 바가 맞을지  확인 부탁드려요! ㅎ

wisemuji reviewed (approved):
5단계 동안 같이 미션 진행하시느라 넘 고생 많으셨습니다 광유님!
리뷰 반영 정말 잘 해주셨습니다 👏

변경해주신 구조와 코멘트를 보니 마지막까지 어떤 구조가 더 좋은 구조일지 열심히 고민해보신 흔적이 보입니다.
이번 미션을 통해 많은걸 얻어가셨길 바랍니다!

남은 미션도 화이팅입니다 💪  앞으로도 응원할게요!!!
